mars/learn/linear_model/_base.py

```diff
@@ -0,0 +1,319 @@
+# Copyright 1999-2021 Alibaba Group Holding Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numbers
+from abc import ABCMeta, abstractmethod
+
+import scipy.sparse as sp
+from numpy.linalg import LinAlgError
+from sklearn.utils.validation import (check_is_fitted,
+                                      _deprecate_positional_args)
+from sklearn.base import MultiOutputMixin
+from sklearn.utils.extmath import safe_sparse_dot
+
+from ... import tensor as mt
+from ..base import BaseEstimator, RegressorMixin
+from ...lib import sparse as mars_sp
+from ..preprocessing import normalize as f_normalize
+from ...tensor.datasource import tensor as astensor
+from ..utils.validation import _check_sample_weight, check_array, FLOAT_DTYPES
+
+
+def _preprocess_data(
+    X,
+    y,
+    fit_intercept,
+    normalize=False,
+    copy=True,
+    sample_weight=None,
+    return_mean=False,
+    check_input=True,
+):
+    """Center and scale data.
+
+    Centers data to have mean zero along axis 0. If fit_intercept=False or if
+    the X is a sparse matrix, no centering is done, but normalization can still
+    be applied. The function returns the statistics necessary to reconstruct
+    the input data, which are X_offset, y_offset, X_scale, such that the output
+
+        X = (X - X_offset) / X_scale
+
+    X_scale is the L2 norm of X - X_offset. If sample_weight is not None,
+    then the weighted mean of X and y is zero, and not the mean itself. If
+    return_mean=True, the mean, eventually weighted, is returned, independently
+    of whether X was centered (option used for optimization with sparse data in
+    coordinate_descend).
+
+    This is here because nearly all linear models will want their data to be
+    centered. This function also systematically makes y consistent with X.dtype
+    """
+    if isinstance(sample_weight, numbers.Number):
+        sample_weight = None
+    if sample_weight is not None:
+        sample_weight = astensor(sample_weight)
+
+    if check_input:
+        X = check_array(X, copy=copy, accept_sparse=['csr', 'csc'],
+                        dtype=FLOAT_DTYPES)
+    elif copy:
+        if sp.issparse(X):
```

Comment:
For Mars object, use X.issparse() to identify if it's sparse.

Reply:
How about using mar.lib.sparse.issparse(X) instead, cuz we need to consider mars objects as well as np objects

Reply:
OK, reasonable.

Reply:
I checked `mars.lib.sparse.issparse`, it can actually not be used for Mars object, lib is sort of underlying library that only recognize the underlying objects instead of the high-level one, thus I think it's not appropriate to use them to recognize all sparse objects. Two solutions, 

1. In learn, all objects have been converted to Mars object, no numpy ndarray or pandas DataFrame, so `X.issparse()` is enough.
2. Write a util function in which both the high-level Mars objects as well as the low-level objects can be recoginized.

Reply:
Yep, I also found there is something tricky with mars.lib.sparse.issparse(X), part of my debugging shown as follows
```
(Pdb) p X.execute()
<mars.lib.sparse.matrix.SparseMatrix object at 0x7fa4ecfc3730>
(Pdb) p mars_sp.issparse(X)
False
(Pdb) p X.issparse()
True
```
Guess it is what you call high-level v.s. low-level objects. For simplicity, I chose solution 1.